### DISCUSSION TITLE
Nuxt 3 - server-side: i18n switches back to default locale after setting locale

### DISCUSSION THREAD

**Discussion by githubnikola** [nuxt/nuxt#21678] [Questions]
Node: 18.16.0
Nuxt: 3.4.3
@nuxtjs/i18n: 8.0.0-beta.7
@pinia/nuxt: 0.4.7
pinia: 2.0.33
--------------------------
Initial request comes with query param ?lang=sr. I want to set i18n to use this locale on server and that locale to persist on client.

i18n.config.mjs
```
import sr from '~/locales/sr_RS.json'
import hr from '~/locales/hr_HR.json'
import sl from '~/locales/sl_SI.json'

export default {
  locales: [
    { code: 'sr', iso: 'sr_RS' },
    { code: 'hr', iso: 'hr_HR' },
    { code: 'sl', iso: 'sl_SI' },
  ],
  locale: 'sr_RS',
  lazy: true,
  defaultLocale: 'sr_RS',
  fallbackLocale: 'sr_RS',
  messages: {
    sr_RS: sr,
    hr_HR: hr,
    sl_SI: sl
  }
}
```

nuxt.config.js

```
export default defineNuxtConfig({
  modules: ['@nuxtjs/i18n'],
  i18n: {
    vueI18n: './config/i18n.config.mjs'
  }
})
```

Now, I'm using plugin that runs only on server which uses store to load some app data among which is setting locale using `await $i18n.setLocale(localeFromCustomCookie)`. Looking at the console using `onBeforeLanguageSwitch` i see that the locale changed but then it switches back to default.

What is the correct way to do this?

Thanks.

**Comment by Kiansa** [comment]
I had similar issue and apparently you have to add the i18n configuration in nuxt config as well as i18n config file.

This is how i18n in my nuxt config file looks like:

```

i18n: {
    experimental: {
      jsTsFormatResource: true,
    },
    precompile: {
      strictMessage: false,
      escapeHtml: true,
    },
    langDir: 'locales',
    lazy: true,
    baseUrl: 'http://localhost:3000',
    locales: [
      {
        code: 'en',
        name: 'English',
        iso: 'en-US',
        file: 'en.json',
      },
      {
        code: 'es',
        name: 'Español',
        iso: 'es-ES',
        file: 'es.json',
      },
      {
        code: 'fr',
        name: 'Français',
        iso: 'fr-FR',
        file: 'fr.json',
      },
    ],
    // trailingSlash: true,
    debug: false,
    defaultLocale: 'en',
    // strategy: 'no_prefix',
    // strategy: 'prefix',
    // strategy: 'prefix_and_default',
    strategy: 'prefix_except_default',
    // rootRedirect: '/ja/about-ja',
    dynamicRouteParams: true,
    // customRoutes: 'config',

    // differentDomains: true,
    // skipSettingLocaleOnNavigate: true,
    // detectBrowserLanguage: false,
    detectBrowserLanguage: {
      useCookie: true,
      // alwaysRedirect: true
      // cookieKey: 'i18n_redirected',
      // // cookieKey: 'my_custom_cookie_name',
      // redirectOn: 'root'
    },
    vueI18n: './i18n.config.ts',
  },

```

You check this starter template for complete setup [Nuxt Ultra](https://t.co/FvroVtw3T9)

**Reply by githubnikola** [reply]
@Kiansa 
Thank you, I'll give it a try tomorrow.

**Reply by githubnikola** [reply]
@Kiansa 
1. I haven't changed any version of listed libraries.
2. I have just added to nuxt.config which now looks like this.
```
i18n: {
  locales: [
    { code: 'sr', iso: 'sr_RS' },
    { code: 'hr', iso: 'hr_HR' },
    { code: 'sl', iso: 'sl_SI' },
  ],
  defaultLocale: 'sr_RS',
  fallbackLocale: 'sr_RS',
  strategy: 'no_prefix',
  messages: {
    sr_RS: sr,
    hr_HR: hr,
    sl_SI: sl
  },
  vueI18n: './config/i18n.config.mjs'
}
```

It does not solve my issue.
I want to call `setLocale(myLocale)` on initial request on the server. After calling it, i18n defaults back to "defaultLocale". I assume that I might be calling the method at the wrong moment, before some internal i18n methods are called, or something else is happening.
Calling it on the client results in showing default locale for a brief moment, and then changing it to the other one. I've seen some threads about that "flicker", but none of the proposed solutions there helped. Also, there were some ssr to client persistence posts, but also did work.

I have checked your repo and what I could see is that you are using different setup. I might be missing something crucial here, but I don't know what.

EDIT: Also I don't want locale to be part of the url.

**Reply by Kiansa** [reply]
Based on the information you provided, I think what you are looking and missing is [this part of the configuration](https://i18n.nuxtjs.org/browser-language-detection) 

To redirect the user every time they visit the app and keep their selected choice, enable alwaysRedirect

```
i18n: {
  // ...
  detectBrowserLanguage: {
    useCookie: true,
    // cookieKey: 'my_custom_cookie_name',  // if you want to set your own cookie name. defaults to i18n_redirected
    alwaysRedirect: true
  }
}
```

I think in this approach, the initial visit checks for the browser language. If it doesn't exist, the default locale is set. If the user changes the locale, locale is saved in a cookie for future visits, ensuring the correct locale is displayed.

**Reply by githubnikola** [reply]
@Kiansa 
Ok, I'll check that ASAP.

**Reply by githubnikola** [reply]
@Kiansa 
I must be doing something wrong here, because this is not working for me. Yes, the cookie is set but it does not have an effect.

**Comment by timothejoubert** [comment]
I made as simple reproduction [here](https://stackblitz.com/edit/nuxt-starter-i18n-timothe-rmrfwdrt?file=pages%2F%5B...slug%5D.vue).
Setting the locale on the server side does not persist on the client.

**Comment by ericnshimiye** [comment]
I know this is a bit late, but I ran into the same issue and wanted to share what worked for me.

The problem in my case was related to how [routing strategies](https://i18n.nuxtjs.org/docs/guide#routing) are configured in @nuxtjs/i18n. By default, the module overrides Nuxt's routing to include locale prefixes in URLs (unless you're using the 'no_prefix' strategy).

This means that setting the locale manually can be overridden by a redirect or route change that resets it based on the URL, depending on the strategy you're using.

Switching to the 'no_prefix' strategy solved it for me, since it doesn't tie the locale to the route path. Here’s the config that worked:

```ts
i18n: {
  locales: [...],
  defaultLocale: 'en',
  strategy: 'no_prefix',
  lazy: true,
  detectBrowserLanguage: {
    useCookie: true,
    alwaysRedirect: true,
    redirectOn: 'root'
  }
}
```

**Reply by hedgehog9898** [reply]
God bless your soul, thank you for the sharing! Works for me 👍
